lib/async_hooks.js

```diff
@@ -342,24 +345,32 @@ function emitInitS(asyncId, type, triggerAsyncId, resource) {
   }
 }
 
-
-function emitBeforeN(asyncId) {
-  processing_hook = true;
-  // Use a single try/catch for all hook to avoid setting up one per iteration.
-  try {
-    for (var i = 0; i < active_hooks_array.length; i++) {
-      if (typeof active_hooks_array[i][before_symbol] === 'function') {
-        active_hooks_array[i][before_symbol](asyncId);
+function emitHookFactory(symbol, name) {
+  // Called from native. The asyncId stack handling is taken care of there
+  // before this is called.
+  const fn = function(asyncId) {
+    processing_hook = true;
+    // Use a single try/catch for all hook to avoid setting up one per
+    // iteration.
+    try {
+      for (var i = 0; i < active_hooks_array.length; i++) {
+        if (typeof active_hooks_array[i][symbol] === 'function') {
+          active_hooks_array[i][symbol](asyncId);
+        }
       }
+    } catch (e) {
+      fatalError(e);
     }
-  } catch (e) {
-    fatalError(e);
-  }
-  processing_hook = false;
+    processing_hook = false;
 
-  if (tmp_active_hooks_array !== null) {
-    restoreTmpHooks();
-  }
+    if (tmp_active_hooks_array !== null) {
+      restoreTmpHooks();
+    }
+  };
+  Object.defineProperty(fn, 'name', {
```

Comment:
Mind leaving a comment that explains this sets the anonymous function name such it looks good in stack traces.